solutions/usaco-994.mdx

```diff
@@ -2,49 +2,134 @@
 id: usaco-994
 source: USACO Gold 2020 January
 title: Farmer John Solves 3SUM
-author: Qi Wang
+author: Qi Wang, Ryan Chou
 ---
 
 [Official Analysis (C++ and Java)](http://www.usaco.org/current/data/sol_threesum_gold_jan20.html)
 
+## Explanation
+
+Although the problem may be daunting at first, it helps to think about how a smaller range is connected to another.
+
+If $\texttt{ways}[i][j]$ equals the number of unordered triplets which sum to zero between $[i, j]$, then what states does this rely on?
+
+![](usaco-994/init.png)
+
+Well intuitively, we can choose to exclude one endpoint or the other, which gives us a picture like this (reminder to make sure not to double count the overlap).
+
+![](usaco-994/ps.png)
+
+While this'll count all the triplets in between them, we aren't taking into account any triplets which use both the elements at $i$ and $j$, since those aren't included in our "PIE-like" transition.
+
+To do this, we can precalculate the number of triplets which start at $i$ and end at $j$ in $\mathcal{O}(N^2)$ time by storing the occurences of the compliment of $i$ and $j$ in an array.
```

Comment:
> which start at $i$ and end at $j$

or just "which use both the $i$th and $j$th elements"

Reply:
hmm, well i wanna make it clear that those are the endpoints

Reply:
but this is an ordered triple? not an interval?

Reply:
yeah so like the triple's first element is i and the third element is j